Download manager dummy PR

aaronlehmann opened the pull request:
Opening this PR for review comments only.

dongluochen committed:
move container files to a separate folder. Following changes will update them.

Signed-off-by: Dong Chen <dongluo.chen@docker.com>

dongluochen committed:
Add container package for container APIs.

Signed-off-by: Dong Chen <dongluo.chen@docker.com>

Commit:
Add download mananger

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Use RootFS instead of Layer for result

Pass in an initial RootFS to handle base layers.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Replace Layer interface method with DiffID

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Remove Registered from main interface. Detect it by casting.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
godoc

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Concurrency limit

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Use the download manager for v1 pulls

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Remove Pool

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Add upload manager; use it for uploads

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aidanhs committed:
Make releaseLayers iterative rather than recursive, remove useless func

Signed-off-by: Aidan Hobson Sayers <aidanhs@cantab.net>

Commit:
v1 pull cleanups

- Remove pull -a concurrency
- Remove unnecessary json unmarshal

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Make download manager's Download a blocking function

Change the interface between the puller and the daemon to send progress
over a channel.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Make the upload manager's Upload a blocking function

Change the interface between the pusher and the daemon to send progress
over a channel.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Use ioutils.NewReadCloserWrapper instead of a custom wrapper

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

unclejack committed:
Dockerfile: add python-dev to fix Python.h errors

Signed-off-by: Cristian Staretu <cristian.staretu@gmail.com>

dmcgowan commented on distribution/pull_v1.go:
```diff
@@ -360,95 +274,57 @@ func (p *v1Puller) downloadLayerConfig(out io.Writer, v1LayerID, endpoint string
-		}
+	defer layerReader.Close()
+
+	tmpFileWrapper := &tmpFileWrapper{}
```
choose different variable name

dmcgowan commented on distribution/pull.go:
```diff
@@ -183,3 +176,25 @@ func validateRepoName(name string) error {
 	return nil
 }
+
+type tmpFileWrapper struct {
```
Based on usage of this wrapper I think it could fully wrap the `tmpFile` by exposing a Write method and providing a constructor which makes the tempfile. This could also protect calling a write after `didFirstRead` is true.

aaronlehmann commented:
cc @tonistiigi

tonistiigi commented on distribution/pull_v1.go:
```diff
@@ -360,95 +274,57 @@ func (p *v1Puller) downloadLayerConfig(out io.Writer, v1LayerID, endpoint string
+	progressChan <- xfer.Progress{ID: ld.ID(), Action: "Download complete"}
+
+	logrus.Debugf("Downloaded %s to tempfile %s", ld.ID(), tmpFileWrapper.tmpFile.Name())
+
```
`ioutils.NewReadCloserWrapper` here should do instead of the `tmpFileWrapper`.

tonistiigi commented on distribution/pull_v2.go:
```diff
@@ -269,119 +248,22 @@ func (p *v2Puller) pullV2Tag(out io.Writer, ref reference.Named) (tagUpdated boo
-		d.broadcaster.Write(p.sf.FormatProgress(stringid.TruncateID(d.digest.String()), "Pull complete", nil))
-		d.broadcaster.Close()
-		tagUpdated = true
+	resultRootFS, release, err := p.config.DownloadManager.Download(context.Background(), *rootFS, descriptors, p.config.ProgressChan)
```
I think its fine to use context for the progressChan here. `pkg.FromContext()`, `pkg.NewContext()` like in https://godoc.org/golang.org/x/net/trace for example.

aaronlehmann replied:
You mean pass the channel as an untyped named argument on the context? Seems ugly IMHO, and I'm not sure what benefit it would provide.

tonistiigi replied:
Its not untyped. `fromContext` returns progress(writer) and `newContext` takes in something that can be read from. I think its cleaner, especially when callbacks etc also use this method.

tonistiigi replied:
I also makes progress optional for the clients.

tonistiigi replied:
cc @stevvooe

Commit:
Implement retries

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

cpuguy83 committed:
Fix race in locker call to `dec()`

Can't safely use uint32 for locker since we need to decrement the count,
which requires loading the unit and doing some math, which is inherintly
racey.
Instead use Int32 which we can safely use with atomic and AddInt32 with
`-1`

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

LK4D4 committed:
Merge pull request #18196 from cpuguy83/18190_on_your_marks

Fix race in locker call to `dec()`

calavera committed:
Move versioned references of inspect functions to the daemon.

Leaving only one versioned main function that a backend must implement.

Signed-off-by: David Calavera <david.calavera@gmail.com>

LK4D4 committed:
Merge pull request #18150 from unclejack/add_python_dev

Dockerfile: add python-dev to fix Python.h errors

calavera committed:
Split container backend into several specialized interfaces.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Merge pull request #17924 from aaronlehmann/content-addressability

Content addressability

LK4D4 committed:
Merge pull request #18203 from aidanhs/aphs-content-addressability-layerstore-releaselayer

Make releaseLayers iterative rather than recursive, remove useless func

cpuguy83 committed:
Merge pull request #18201 from calavera/split_container_router

[Carry 17736] Container router and backend.

Commit:
Implement cancellation

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Closing in favor of https://github.com/aaronlehmann/docker/pull/6
